lib/galaxy/datatypes/binary.py

```diff
@@ -981,78 +981,81 @@ def sniff(self, filename):
-                    dataset.metadata.var_size = _layercountsize(tmp, len(tmp["index"]))
-
-            if 'varm' in dataset.metadata.layers_names:
-                tmp = anndata_file["varm"]
-                dataset.metadata.varm_layers, dataset.metadata.varm_count, _ = _layercountsize(tmp)
-
-            if 'uns' in dataset.metadata.layers_names:
-                tmp = anndata_file["uns"]
-                dataset.metadata.uns_layers, dataset.metadata.uns_count, _ = _layercountsize(tmp)
-
-            # Resolving the problematic shape parameter
-            if 'X' in dataset.metadata.layers_names:
-                # Shape we determine here due to the non-standard representation of 'X' dimensions
-                shape = anndata_file['X'].attrs.get("shape")
-                if shape is not None:
-                    dataset.metadata.shape = tuple(shape)
-                elif hasattr(anndata_file['X'], 'shape'):
-                    dataset.metadata.shape = tuple(anndata_file['X'].shape)
-
-            if dataset.metadata.shape is None:
-                dataset.metadata.shape = (int(dataset.metadata.obs_size), int(dataset.metadata.var_size))
+        try:
+            with h5py.File(dataset.file_name, 'r') as anndata_file:
+                dataset.metadata.title = util.unicodify(anndata_file.attrs.get('title'))
+                dataset.metadata.description = util.unicodify(anndata_file.attrs.get('description'))
+                dataset.metadata.url = util.unicodify(anndata_file.attrs.get('url'))
+                dataset.metadata.doi = util.unicodify(anndata_file.attrs.get('doi'))
+                dataset.creation_date = util.unicodify(anndata_file.attrs.get('creation_date'))
+                dataset.metadata.shape = anndata_file.attrs.get('shape') or dataset.metadata.shape
+                # none of the above appear to work in any dataset tested, but could be useful for future
+                # AnnData datasets
+
+                # all possible keys
+                dataset.metadata.layers_count = len(anndata_file)
+                dataset.metadata.layers_names = list(anndata_file.keys())
+
+                def _layercountsize(tmp, lennames=0):
+                    "From TMP and LENNAMES, return layers, their number, and the length of one of the layers (all equal)."
+                    if hasattr(tmp, 'dtype'):
+                        layers = [util.unicodify(x) for x in tmp.dtype.names]
+                        count = len(tmp.dtype)
+                        size = int(tmp.size)
+                    else:
+                        layers = [util.unicodify(x) for x in list(tmp.keys())]
+                        count = len(layers)
+                        size = lennames
+                    return (layers, count, size)
+
+                if 'obs' in dataset.metadata.layers_names:
+                    tmp = anndata_file["obs"]
+                    dataset.metadata.obs_names = [util.unicodify(x) for x in tmp["index"]]
```

Comment:
```suggestion
                    tmp = anndata_file["obs"]
                    if "index" in tmp:
                        obs_index = "index"
                    else:
                        obs_index = "_index"
                    dataset.metadata.obs_names = [util.unicodify(x) for x in tmp[obs_index]]
```
Maybe we can take the chance to cater for the "_index" 0.7.4 case

Reply:
feel free to neglect as per below comments.